### ISSUE TITLE
Bolt asset versioning strategy differers from symfony

### ISSUE THREAD

**Issue by cdowdy** [bolt/bolt#7483]
Using the asset twig function we can't turn off the query string, or the config parameters are ignored since the bolt docs and symfony docs differ

Details
-------

 - Relevant Bolt Version: [ 3.5.1 ]
 - Install type: [ Composer install  ]
 - PHP version: [ 7.2.4-1 ]
 - Used web server: [ Nginx  version 1.13.12

Reproduction
------------

### Bug summary

Following the bolt docs on linking to assets in templates https://docs.bolt.cm/3.5/templating/linking-in-templates#using-asset-to-link-to-assets-or-files and then following the link to Symfony's docs on the asset function and version _strategy http://symfony.com/doc/current/reference/configuration/framework.html#version-strategy  we can't turn off the query string or use a json manifest. 

### Steps to reproduce

include an asset using the asset function:  
  * ``<script src="{{ asset('dist/app.js', 'theme') }}"></script>``

include the symfony yaml snippet to change the version strategy or use the json_mainifest option: 

no version:  
```yaml   
framework:
     assets:
         packages:
             theme:
                 # this package removes any versioning (its assets won't be versioned)
                 version: ~
```  

json_mainifest:   

```yaml   
framework:
     assets:
         packages:
             theme:
                 # this package uses its own manifest (the default file is ignored)
                 json_manifest_path: '/dist/mainfest.json'
```  

refresh page the assets are used on

### Expected result

asset will be delivered with no query string like or use the json manifest like so:  

``<script src="/theme/myTheme/dist/app.js"></script>`` 

### Actual result

the assets are still served with bolt's version strategy or ignores the json manifest option.  

------

The reason this happens is bolt uses a different way to get the configs for the asset fuction.  
  
Symfony uses the above examples and bolt uses   

```yaml  
assets:
   theme:
        json_manifest_path: "/dist/manifest.json"
``` 

The ``framework``,  and ``packages`` config options are not present in the way bolt applies the versioning strategy, and is not mentioned in the docs that they differ and since the symfony docs are linked too one would assume that bolt follows the same conventions.  

I noticed this after I was linked to the asset provider https://github.com/bolt/bolt/blob/master/src/Provider/AssetServiceProvider.php#L60 on slack

**Comment by GwendolenLynch** [comment]
They are handled differently and use Symfony's DI to do several things. Bolt's is in v3 hooked into some legacy stuff.

The good news is that if you want to build your own service, you can reinject that into the application container and override Bolt's.

**Comment by cdowdy** [comment]
so in order to use webpack to build our assets for themes we should build a new service? or are we talking about two different things haha?

**Comment by GwendolenLynch** [comment]
To use Webpack versioning, yes, unfortunately … I looked at what SF had done, but porting it to v3 was way more work that I have time for.

FWIW I use Webpack for all my asset builds on Bolt sites these days, it just means using Bolt's asset versioning.

**Comment by cdowdy** [comment]
Ok I think we are talking about different things or I'm just being dense after work today haha

My issue is about bolt using an undocumented configuration for the asset versioning strategy. Bolts docs link directly to symfonys docs. 

If you use symfony configuration nothing works. If you crawl through bolts code you'll see bolt doesn't use the same configuration pattern. Once you use the undocumented bolt configuration you can use the asset function with either no query string or json manifest.

Using bolts built in asset versioning won't work for service workers since we have no control over the query string when using webpack to build our service worker

**Comment by GwendolenLynch** [comment]
We might be … It's been a long and stressful few weeks … I'll have a reread in the morning when I've had some sleep, mate :smile_cat:

**Comment by cdowdy** [comment]
Haha no worries get back to it when you can!

**Comment by cdowdy** [comment]
To add to this I was also developing a theme and wanted to use webpack's hot module replacement (coupled with webpack dev server ) and the `base_urls` / `base_url` options - as seen here http://symfony.com/doc/current/reference/configuration/framework.html#packages - aren't implemented.  

just to give a little info on webpack-dev-server / hmr and a bolt we could use the  `config_local.yml` file and in a development mode - passed in by a config switch or however you wish to do it -  and supply all the paths and version strategy. This could look like:  
  

```yaml  
assets:
   theme:
        json_manifest_path: "/dist/manifest.json"  
        base_url: ''http://localhost:8080/'
```  

Then when we develop a theme all the assets we have in `/dist` be it css/js etc will be served from webpacks dev server and we'd have Hot Module Replacement, similar to livereload/browsersync. 

The reason this is an "issue" is that bolt's docs link to symfonys docs on asset function   

>For a more in-depth description of the asset function, see the Symfony documentation on assets.  

So one would assume that the things mentioned on symfonys page(s) are also applicable to bolt since bolt just passes the rest of the docs to symfony. 

They are in some instances but at the same time are not applicable to bolt since bolt doesn't follow the same configuration setup/patterns as the linked symfony docs and hasn't implemented some of the "features" offered by the asset function (well from a quick glance at https://github.com/bolt/bolt/tree/master/src/Asset) I could be way off though haha.. 


 

to summarize this basically comes down to a documentation thing I guess.  

We could update the docs with examples to show that bolt only offers the version strategy, and json_manifest options with the appropriate configs that bolt uses instead of the linked symfony docs and examples. If that's the route you all agree on I'll get a PR up in a bit

**Comment by GwendolenLynch** [comment]
OK, still not seeing your point here.

Our docs link to [Symfony's templating page](http://symfony.com/doc/current/templating.html#templating-assets), whereas what you reference is for Symfony Framework's asset bundle configuration. We've never used the framework within Bolt.

This is coming in v4, but prior to then I stand by my first comment here.

**Comment by cdowdy** [comment]
> Our docs link to Symfony's templating page

Which symfonys docs links to the framework bundle on asset versioning.

If you all didn't use the symfony version then why is there a  bolt json manifest version strategy and version strategy provider? 

If the bolt asset function is a complete custom one off thing then why even link to symfonys docs on the asset function. You can't do anything mentioned in the symfony docs on how to version assets or customize the asset function. 

If you're telling me in order for a front end developer to use webpack like they can for any other CMS in bolt is to write a custom provider (even though it's set up just doesn't follow the symfony asset configuration standard)  then I'll say ok and move along haha

**Comment by GwendolenLynch** [comment]
> Which symfonys docs links to the framework bundle on asset versioning.

Care Factor Zero … They also link to the rest of their docs, that doesn't mean we implement the whole damn framework. That is rather an insulting comment btw.

**Comment by cdowdy** [comment]
>Care Factor Zero … They also link to the rest of their docs, that doesn't mean we implement the whole damn framework. That is rather an insulting comment btw. 

so me.. stating a fact that the docs bolt passes off to symfony and then symfony passes the to the symfony framework edition is insulting? Well I apologize then for whatever slight that may have caused.

and me. offering to fill in the gaps of the documentation, document bolts undocumented use of something is care factor zero? alright then

**Comment by cdowdy** [comment]
heck you yourself @GawainLynch did the json manifest pull request with the differing config https://github.com/bolt/bolt/pull/7305  

I was asking if we could update the docs with appropriate, actually documented examples. But since I had  go through symfonys docs - which bolt links to for "For a more in-depth description of the asset function, see the Symfony documentation on assets." - I figured the same config would be used which is not the case and didn't see the pull which is the entire reason I filed this issue. There's no documentation on how to use json_manifest. ONe would have to find your pull and see you used a different configuration and mount point. 

instead when I offered to contribute to the project I was met with   

> Care Factor Zero …  

and told I was being insulting by stating an indisputable fact. Really the only thing I see or find insulting here is the way you closed it and told me you didn't care.    

Good luck. Sorry for wanting to contribute. Maybe I'll mark the rest of my extensions and boilerplates as no longer maintained if this is how things are going to be.

**Comment by rixbeck** [comment]
:disappointed:

**Comment by bobdenotter** [comment]
@cdowdy Hey Cory, hop on the Bolt Slack, if you will.. I'd like to have a brief chat with you.

**Comment by bobdenotter** [comment]
Fixed in bolt/docs#944.

@cdowdy You might like to hear that today we landed Symfony Encore 0.21 in Bolt Four, with Webpack 4 and Vue Loader 15. :-D
